src/VegaWrapper2.js

```diff
@@ -0,0 +1,403 @@
+        this.loader.file = () => { throw new Error('Disabled'); };
+    }
+
+    /**
+     * Validate and update urlObj to be safe for client-side and server-side usage
+     * @param {object} uri - An object that will be converted into an url string
+     * @param {object} options - passed by the vega loader
+     * @return {Promise} The sanitized url is provided by the 'href' property. We never load file from local file system.
+     */
+    sanitize(uri, options) {
+        return Promise.resolve({href: this.objToUrl(uri, options), loadFile: false});
+    }
+
+    /**
+     * Check if host was listed in the allowed domains, normalize it, and get correct protocol
+     * @param {string} host
+     * @returns {Object}
+     */
+    sanitizeHost(host) {
+        // First, map the host
+        host = (this.domainMap && this.domainMap[host]) || host;
+
+        if (this.testHost('https', host)) {
+            return { host: host, protocol: 'https' };
+        } else if (this.testHost('http', host)) {
+            return { host: host, protocol: 'http' };
+        }
+        return undefined;
+    };
+
+    /**
+     * Test host against the list of allowed domains based on the protocol
+     * @param {string} protocol
+     * @param {string} host
+     * @returns {boolean}
+     */
+    testHost(protocol, host) {
+        if (!this.validators[protocol]) {
+            const domains = this.domains[protocol];
+            if (domains) {
+                this.validators[protocol] = makeValidator(domains, protocol === 'https' || protocol === 'http');
+            } else {
+                return false;
+            }
+        }
+        return this.validators[protocol].test(host);
+    };
+
+    /**
+     * Override the protocol and host for *wikidatasparql*, *geoshape*, *geoline* and *mapsnapshot*
+     * @param {object} urlParts output the new result to this object
+     * @param {object} urlObj used to log and determine the corresponding host
+     * @param {string} protocolOverride used to determine the corresponding host
+     * @private
+     */
+    _overrideHostAndProtocol(urlParts, urlObj, protocolOverride) {
+        let protocol = protocolOverride || urlObj.type,
+            domains = this.domains[protocol];
+        if (!domains) {
+            throw new Error(protocol + ': protocol is disabled: ' + JSON.stringify(urlObj));
+        }
+        urlParts.host = domains[0];
+        urlParts.protocol = this.sanitizeHost(urlParts.host).protocol;
+    }
+
+    /**
+     * convert the urlObj to a url string
+     * @param {object} urlObj an object consists of type and essential parameters
+     * @param {object} options used to attach CORS infomation
+     * @returns {string} a complete url
+     */
+    objToUrl(urlObj, options) {
+        const sanitizedHost = this.sanitizeHost(urlObj.wiki ? urlObj.wiki : options.domain);
+
+        if (!sanitizedHost) {
+            throw new Error('URL hostname is not whitelisted: ' + urlParts.host);
```

Comment:
`urlParts.host` does not exist here
```suggestion
            throw new Error('URL hostname is not whitelisted: ' + host);
```

Reply:
You did not fix the issue here. Please don't mark conversation resolved until you actually add a fix.  Same for some other issues.  Please recheck all my comments and address them.

Reply:
I will finish it soon and double check your comments.